feat: simplify notifications about background task in admin

h1k3r opened the pull request:
### 1. Why is this change necessary?

Please check original issue.

### 2. What does this change do, exactly?

- introduces `SentAtStamp` and middleware that adds this stamp to every message
- adds message subscriber to store statistics on processing
- introduces new table `messenger_stats`, repository and abstractions to support other types of storages
- adds new module in administration/settings to display statistics

### 3. Describe each step to reproduce the issue or behaviour.
- execute operations that will trigger messages creation/processing
- navigate in administration to Settings -> System -> Message queue statistics

### 4. Please link to the relevant issues (if any).
- closes https://github.com/shopware/shopware/issues/8697
- relates https://github.com/shopware/shopware/issues/4121

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

h1k3r committed:
NEXT-36569 - implemented simplified messenger statistics

h1k3r committed:
NEXT-36569 - implemented simplified mysql based messenger statistics

h1k3r committed:
NEXT-36569 - smaller fixes, removed circular buffer based stats

h1k3r committed:
NEXT-36569 - migration test

h1k3r committed:
NEXT-36569 - unit tests

h1k3r committed:
NEXT-36569 - integration tests, introduced entity objects, changed API format, changes db schema

h1k3r committed:
NEXT-36569 - renamed entities

h1k3r committed:
NEXT-36569 - added messenger_stats table to definition validator

h1k3r committed:
NEXT-36569 - added OpenAPI schema

h1k3r committed:
NEXT-36569 - preparation to support redis storage

h1k3r committed:
NEXT-36569 - test improvements

h1k3r committed:
NEXT-36569 - fixed mr comments

h1k3r committed:
NEXT-36569 - added needed index in migration

h1k3r committed:
NEXT-38422 - php-cs-fixer updated ruleset run

h1k3r committed:
Basic admin module implementation for message stats

h1k3r committed:
-updated translation

h1k3r committed:
-made banner possible to hide

mitelg commented on src/Core/Framework/MessageQueue/Stats/Entity/MessageStatsEntity.php:
```diff
@@ -0,0 +1,50 @@
+    protected MessageTypeStatsCollection $messageTypeStats;
+
+    public function __construct(
+        int $totalMessagesProcessed,
```
why no promoted properties?

h1k3r replied:
Thanks, old code :),  updated

h1k3r committed:
-added tooltips

h1k3r committed:
-added styling

h1k3r committed:
-added loading real statistics

h1k3r committed:
-php pipeline fixes

h1k3r committed:
-php pipeline fixes - coverage

h1k3r committed:
-js pipeline fixes - lint

h1k3r committed:
-js pipeline fixes - added tests, updated specs

h1k3r committed:
-js pipeline fixes - added missing test

h1k3r commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/index.spec.js:
```diff
@@ -0,0 +1,60 @@
+
+const { Module } = Shopware;
+
+describe('src/module/sw-settings-message-stats/index.js', () => {
```
I am not sure if such tests make sense, added it as found similar in the `sw-settings-media` module.

jleifeld replied:
This test makes sense, is correct 👍

h1k3r committed:
-switch to promoted properties

h1k3r committed:
-added changelog

h1k3r commented on src/Core/Framework/MessageQueue/Stats/README.md:
todo: remove before merge

@mitelg example of module readme we were talking about (but probably administration readme should be in a separate folder).

mitelg replied:
Looks good :ok_hand: 

we should think about adding those

mitelg commented on src/Core/Framework/Api/Controller/InfoController.php:
```diff
@@ -103,6 +105,12 @@ public function queue(): JsonResponse
         ], array_values($entries)));
     }
 
+    #[Route(path: '/api/_info/message-stats.json', name: 'api.info.queue2', methods: ['GET'])]
```
can we have something different for the name? :thinking:

h1k3r replied:
Changed

mitelg commented on src/Core/Framework/MessageQueue/Stats/StatsService.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function registerMessage(Envelope $envelope): void
+    {
+        $sentAtStamp = $envelope->last(SentAtStamp::class);
```
in the middleware we also consider `SentStamp`. would it be possible, that we miss some messages in the statistic then? :thinking:

h1k3r replied:
I switched for more clear `ReceivedStamp` to identify that middleware is not running in the receiving stage.

mitelg commented on tests/integration/Core/Framework/Api/Controller/InfoControllerTest.php:
```diff
@@ -670,6 +675,39 @@ public function testFetchApiRoutes(): void
+        static::assertArrayHasKey('totalMessagesProcessed', $stats);
+        static::assertGreaterThanOrEqual(2, $stats['totalMessagesProcessed']);
+        static::assertArrayHasKey('processedSince', $stats);
+        static::assertInstanceOf(\DateTimeInterface::class, \DateTimeImmutable::createFromFormat(\DateTimeInterface::RFC3339_EXTENDED, $stats['processedSince']));
```
can you not just put the value into the constructor of DateTime? :thinking:

h1k3r replied:
It's a lazy check that format of the field is correct: [createFromFormat](https://www.php.net/manual/en/datetimeimmutable.createfromformat.php) returns false otherwise

mitelg commented on tests/integration/Core/Framework/Api/Controller/InfoControllerTest.php:
```diff
@@ -670,6 +675,39 @@ public function testFetchApiRoutes(): void
+        static::assertIsFloat($stats['averageTimeInQueue']);
+        static::assertArrayHasKey('messageTypeStats', $stats);
+        static::assertIsArray($stats['messageTypeStats']);
+        static::assertArrayHasKey('type', $stats['messageTypeStats'][0]);
```
we should also test the contents of the first element

mitelg commented on tests/integration/Core/Framework/MessageQueue/Stats/MySQLStatsRepositoryTest.php:
```diff
@@ -0,0 +1,119 @@
+    {
+        parent::setUp();
+
+        $this->connection = $this->getContainer()->get('Doctrine\DBAL\Connection');
```
why not the class reference?

mitelg commented on tests/integration/Core/Framework/MessageQueue/Stats/MySQLStatsRepositoryTest.php:
```diff
@@ -0,0 +1,119 @@
+        $repository->setNow($this->dateTimeFromTime($now));
+        $repository->updateMessageStats('myclassname', 0);
+
+        static::assertEquals(2, $this->countRecords($this->dateTimeFromTime($now - 21)));
```
please use `assertSame` where possible instead, as it is type safe

h1k3r replied:
Nice that now phpstan also says this! 🥳

mitelg commented on tests/integration/Core/Framework/MessageQueue/Stats/MySQLStatsRepositoryTest.php:
```diff
@@ -0,0 +1,119 @@
+        $repository = new MySQLStatsRepositoryTestable($this->connection, 20);
+
+        $now = 1726671956;
+        $repository->setNow($this->dateTimeFromTime($now - 21));
```
instead of magic numbers, using some constants would make the purpose more clearer. 

e.g. `TOO_OLD` or something like that

mitelg commented on tests/migration/Core/V6_6/Migration1726135997CreateMessengerStatsTableTest.php:
```diff
@@ -0,0 +1,53 @@
+        $this->rollback();
+
+        $migration = new Migration1726135997CreateMessengerStatsTable();
+        $migration->update($this->connection);
```
please execute the update method twice to test multiple executions of the migration

mitelg commented on tests/unit/Core/Framework/Adapter/Messenger/Stamp/SentAtStampTest.php:
```diff
@@ -0,0 +1,24 @@
+    public function testGetSentAt(): void
+    {
+        $sentAt = new \DateTimeImmutable('@123456789');
+        $stamp = new SentAtStamp($sentAt);
```
do we really need to test the stamp?

h1k3r replied:
It has default time creation, so I would keep the test

h1k3r committed:
-removed banner

h1k3r committed:
-styling update to follow the figma design (p1)

h1k3r committed:
-changed rendering of the empty data

h1k3r committed:
-updated rendering of empty state, added proper error notifications

h1k3r committed:
-moved migrations to 6_7 folder, fixing pr comments

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/sw-settings-message-stats.scss:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    // Custom styles for the empty state in this module. Should be removed as soon as sw-empty-state is replaced by mt-empty-state meteor component.
+    .sw-empty-state__element {
```
You can use already the mt-empty-state. Just add it to this list:
```
src/Administration/Resources/app/administration/src/app/adapter/view/vue.adapter.ts
```

and then it is available

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/index.js:
```diff
@@ -0,0 +1,122 @@
+const { Mixin } = Shopware;
+
+// eslint-disable-next-line sw-deprecation-rules/private-feature-declarations
+export default {
```
Would be good if you write everything in Typescript instead of Javascript

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/init/services.init.js:
```diff
@@ -0,0 +1,12 @@
+import MessageStatsApiService from '../../../core/service/api/message-stats.api.service';
+
+// eslint-disable-next-line sw-deprecation-rules/private-feature-declarations
+Shopware.Service().register('messageStatsService', () => {
```
Would be good if you write everything in Typescript instead of Javascript

jleifeld commented on src/Administration/Resources/app/administration/src/core/service/api/message-stats.api.service.js:
```diff
@@ -0,0 +1,37 @@
+ * @class
+ * @extends ApiService
+ */
+class MessageStatsApiService extends ApiService {
```
Would be good if you write everything in Typescript instead of Javascript

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/index.js:
```diff
@@ -0,0 +1,122 @@
+            } catch (error) {
+                this.createNotificationError({
+                    title: this.$tc('sw-settings-message-stats.general.errorTitle'),
+                    message: error.message,
```
Use optional chaining here: `message: error?.message`. You never know if the `error` variable is defined and has the property `message`

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/sw-settings-message-stats.html.twig:
```diff
@@ -0,0 +1,120 @@
+                        </sw-data-grid>
+                    </div>
+
+                    {# element is styled to look lime mt-empty-state, don't forget to remove styles on update #}
```
Use mt-empty-state dircetly

h1k3r replied:
Done. The only thing - it does not support button slots, and default icons differ from design, so I added button below it (waiting for https://github.com/shopware/meteor/pull/730 to be merged)

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/sw-settings-message-stats.scss:
```diff
@@ -0,0 +1,156 @@
+@import "~scss/variables";
+
+.sw-settings-message-stats {
```
CSS looks good just from reading it. I didn't checkout the PR to see if it is looking correct. It would be good if a QA can have a test on this if you're unsure

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/sw-settings-message-stats.spec.js:
```diff
@@ -0,0 +1,76 @@
+    });
+}
+
+describe('module/sw-settings-message-stats/page/sw-settings-message-stats', () => {
```
For me there are missing functionality parts which should be tested. In theory the test should fail if I remove something relevant from the template (or script part) of the component. Now I can remove parts without any failing test but the feature would be broken

jleifeld commented on src/Administration/Resources/app/administration/src/module/sw-settings-message-stats/page/sw-settings-message-stats/sw-settings-message-stats.spec.js:
```diff
@@ -0,0 +1,76 @@
+
+        wrapper.vm.messageStatsService.getStats.mockResolvedValueOnce(mockStats);
+
+        await wrapper.vm.loadStats();
```
Try to avoid accessing `wrapper.vm`. This is only a bandaid solution. Try to use real interfaces like the DOM, props, etc. Here is a detailed description how Jest tests should be written: https://developer.shopware.com/docs/guides/plugins/plugins/testing/jest-admin.html#jest-unit-tests-in-shopware-s-administration

h1k3r committed:
-enabled mt-empty-space component, switched to using it

h1k3r committed:
-regenerated position-identifiers.json

h1k3r committed:
-switched to checking ReceivedStamp to identify if middleware is in sending context

h1k3r committed:
-switched to TypeScript

h1k3r committed:
Merge branch 'trunk' into background-task-notification-mysql

h1k3r committed:
-pipeline fixes

h1k3r committed:
-better test

h1k3r committed:
-fix eslint injected service typing issue

h1k3r committed:
-changed api structure to expose configuration status

h1k3r committed:
-updated admin code for the new api response structure

h1k3r committed:
Merge branch 'trunk' into background-task-notification-mysql

h1k3r committed:
-openapi schema fix

h1k3r committed:
-renaming response entity for consistent naming

h1k3r committed:
Merge branch 'trunk' into background-task-notification-mysql

h1k3r committed:
-added issue link for future improvement

h1k3r committed:
-added test for SentAtStamp default parameter

h1k3r committed:
Internalized remaining classes from stats implementation

Message stats should be available only through API at the moment. Classes can become public apis later if requested

nfortier-shopware reviewed (approved):
👍
